enhancements/microshift/etcd-as-transient-systemd-unit.md

```diff
@@ -0,0 +1,123 @@
+
+2. "As an admin, I want to observe and debug the etcd server logs separately from other microshift processes."
+
+### Goals
+
+- The etcd binary is built and shipped separately from the microshift binary to ease maintainence
+- The etcd server logs are easier to observe
+
+### Non-Goals
+
+- 
+
+## Proposal
+    
+#### Running etcd as a transient systemd unit
+In order to address the concerns brought up in the Motivation section, the MicroShift binary will need to be updated to remove the currently bundled etcd server and replaced with a forked process running under systemd; using the `systemd-run` command. Only the execution of etcd will change - from goroutine to external process - MicroShift's general usage of etcd will not be changed.
+    
+Additionally, this external process will need to be shutdown when MicroShift exits. In an ungraceful termination of MicroShift, it is possible that the etcd systemd unit will continue execution, so it may be necessary to add code to detect this on MicroShift startup so that another etcd launch is not attempted: this would be a port conflict and cause the second etcd instance to fail to start. Further invesigation is needed to learn if a transient systemd unit's lifecycle can be tied to another non-transient systemd unit (MicroShift) so that if the latter exits ungracefully, the former will be gracefully shutdown.
+
+This etcd systemd unit will also need to be provided with the necessary CA certs and signed key pairs for both server and client authentication. 
+    
+The actual execution of etcd in the transient systemd unit will also be considered in this enhancement; there are currently two scenarios (in detail below): running the etcd container image under podman, and running the etcd binary installed via an RPM.
+    
+##### etcd Container Image with Podman
+This scenario would run the etcd server as a container executed via a `podman` command.
+    
+###### Pros
+* This would be similar to how etcd is built and shipped for OCP and would reuse the existing build machinery for openshift/etcd.
+    
+###### Cons
+* The current etcd image is too large, ~400MB, either it will need to be shrunk or a new image will need to be created with the bare minimum in it; the etcd binary alone is about 30MB.
+* This scenario places a runtime dependency on podman; it would need to be installed and ready before MicroShift could be used.
+    
+##### etcd Binary from an RPM
+
+###### Pros
+* This scenario should incur the least amount of non-etcd-related overhead since the binary is being run directly.
+    
+###### Cons
+* New plumbing will need to be created to (or potentially existing will need to be updated/refurbished to) build etcd and package it into an RPM.
+
+### Workflow Description
+
+TODO
+
+#### Variation [optional]
+
+TODO
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+TODO
+### Risks and Mitigations
+
+TODO
+
+### Drawbacks
+
+TODO
+
+
+## Design Details
```

Comment:
It would also be good to understand how we will get an etcd binary upstream. Is that as simple as building the same thing we build for OCP, or would we need anything different?

Reply:
Since etcd has historically been built with an older version of golang, we will also need to work out what to do about golang version in the build. We have the ability to use different golang in container builds, but not RPM builds.